garage/envs/base.py

```diff
@@ -1,17 +1,132 @@
+"""Wrapper class that converts gym.Env into GarageEnv."""
 import collections
 
-from cached_property import cached_property
+import gym
+from gym.spaces import Box as GymBox
+from gym.spaces import Discrete as GymDiscrete
+from gym.spaces import Tuple as GymTuple
+
+from garage.core import Parameterized
+from garage.core import Serializable
+from garage.spaces import Box
+from garage.spaces import Discrete
+from garage.spaces import Product
+
+
+class GarageEnv(gym.Wrapper, Parameterized, Serializable):
+    """
+    Returns a Garage wrapper class for gym.Env.
+
+    In order to provide pickling (serialization) and parameterization
+    for gym.Envs, they must be wrapped with GarageEnv. This ensures
+    compatibility with existing samplers and checkpointing when the
+    envs are passed internally around garage.
+
+    Furthermore, GarageEnv silently converts action_space and
+    observation_space from gym.Spaces to garage.spaces.
+
+    Args:
+        env (gym.Env): the env that will be wrapped
+    """
+
+    def __init__(self, env):
+        super(GarageEnv, self).__init__(env)
+        Parameterized.__init__(self)
+        Serializable.quick_init(self, locals())
+        self._wrapped_env = env
+        self.action_space = self.env.action_space
+        self.observation_space = self.env.observation_space
+        self.reward_range = self.env.reward_range
+        self.metadata = self.env.metadata
+
+    def close(self):
+        """
+        Close the wrapped env.
+
+        Returns:
+            None
+        """
+        self._wrapped_env.close()
+
+    def get_params_internal(self, **tags):
+        """
+        Returns an empty list if env.get_params() is called.
+
+        Returns:
+            An empty list
+        """
+        return []
+
+    @property
+    def horizon(self):
```

Comment:
Isn't this part of the gym interface?

Reply:
`horizon` is not

Reply:
https://github.com/openai/gym/search?q=horizon&unscoped_q=horizon

Reply:
Are you guaranteed that `self._wrapped_env.spec.tags['wrapper_config.TimeLimit.max_episode_steps']` exists for every env we will ever wrap?

Also, that tag is deprecated: https://github.com/openai/gym/blob/master/gym/envs/registration.py#L48

Why not just replace all calls to horizon with env.spec.max_episode_steps?
```python
>>> e = gym.make('Pendulum-v0')
WARN: gym.spaces.Box autodetected dtype as <class 'numpy.float32'>. Please provide explicit dtype.
WARN: gym.spaces.Box autodetected dtype as <class 'numpy.float32'>. Please provide explicit dtype.
>>> dir(e)
['__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_elapsed_seconds', '_elapsed_steps', '_episode_started_at', '_max_episode_seconds', '_max_episode_steps', '_past_limit', '_warn_double_wrap', 'action_space', 'class_name', 'close', 'compute_reward', 'env', 'metadata', 'observation_space', 'render', 'reset', 'reward_range', 'seed', 'spec', 'step', 'unwrapped']
>>> e.spec
EnvSpec(Pendulum-v0)
>>> dir(e.spec)
['__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_entry_point', '_env_name', '_kwargs', '_local_only', 'id', 'make', 'max_episode_seconds', 'max_episode_steps', 'nondeterministic', 'reward_threshold', 'tags', 'timestep_limit', 'trials']
>>>
```

Reply:
Ahh, not all envs have a spec.

Reply:
But that just means we should either first check whether env has a spec inside `horizon`, or just replace all uses of horizon with `env.spec.max_episode_steps`

Reply:
in case 2, what do we do if the env doesn't have a spec?

Reply:
https://github.com/openai/gym/issues/463#issuecomment-389873434

Looks like I can just return `env._max_episode_steps`

Reply:
uhh no you can't because it's _-private

Reply:
also:
```python
(garage36) rjulian@nyquist:~/code/garage$ python
Python 3.6.6 | packaged by conda-forge | (default, Jul 26 2018, 09:53:17)
[GCC 4.8.2 20140120 (Red Hat 4.8.2-15)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from garage.envs.box2d import MountainCarEnv
pygame 1.9.4
Hello from the pygame community. https://www.pygame.org/contribute.html
>>> m = MountainCarEnv()
>>> m._max_episode_steps
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'MountainCarEnv' object has no attribute '_max_episode_steps'
>>>
```

Reply:
Then I'll raise a NotImplementedError within horizon, so that if someone decides to use MountainCarEnv (for instance), `env.horizon` is not a valid call

Reply:
according to my ag, no environments even define horizon

```sh
(garage36) rjulian@nyquist:~/.../garage/garage/envs$ ag horizon
mujoco/randomization/randomized_env.py
77:    def horizon(self):
78:        return self._wrapped_env.horizon

util.py
13:    'bounds', 'default_value', 'flat_dim', 'flatten', 'flatten_n', 'horizon',
74:def horizon(env):

proxy_env.py
36:    def horizon(self):
37:        return self._wrapped_env.horizon

box2d/mountain_car_env.py
17:    @autoargs.arg("goal_cart_pos", type=float, help="Goal horizontal position")
```

Reply:
I'll address the horizon issue in another PR. I'm not sure how much refactoring will be necessary.

Reply:
No environment seems to even define horizon.


```python
scripts/run_experiment.py:103:        help='Print only the tabular log information (in a horizontal format)')
garage/theano/envs/base.py:48:    def horizon(self):
garage/sampler/parallel_sampler.py:116:    :param max_path_length: horizon / maximum length of a single trajectory
garage/sampler/utils.py:164:        # self.horizon_terminals[self.top] = horizon_terminal
garage/sampler/utils.py:263:            # do not pick samples which terminated because of horizon
garage/sampler/utils.py:264:            # if np.any(self.horizon_terminals.take(initial_indices[0:-1],
garage/sampler/utils.py:265:            #    mode='wrap')) or self.horizon_terminals[end_index]:
garage/misc/tabulate.py:60:    """Return a segment of a horizontal line with optional colons which
garage/misc/tabulate.py:74:    """Return a horizontal line with optional colons to indicate column's
garage/misc/tabulate.py:855:    "Return a string which represents a horizontal line."
garage/tf/core/layers.py:1233:                 horizon=None,
garage/tf/core/layers.py:1241:        self.horizon = horizon
garage/tf/core/layers.py:1284:            if self.horizon is not None:
garage/tf/core/layers.py:1286:                for idx in range(self.horizon):
garage/tf/core/layers.py:1776:                 horizon=None,
garage/tf/core/layers.py:1790:        self.horizon = horizon
garage/tf/core/layers.py:1851:            if self.horizon is not None:
garage/tf/core/layers.py:1853:                for idx in range(self.horizon):
garage/tf/algos/batch_polopt.py:36:                 fixed_horizon=False,
garage/tf/algos/batch_polopt.py:81:        self.fixed_horizon = fixed_horizon
garage/envs/mujoco/randomization/randomized_env.py:77:    def horizon(self):
garage/envs/mujoco/randomization/randomized_env.py:78:        return self._wrapped_env.horizon
garage/envs/util.py:13:    'bounds', 'default_value', 'flat_dim', 'flatten', 'flatten_n', 'horizon',
garage/envs/util.py:74:def horizon(env):
garage/envs/proxy_env.py:36:    def horizon(self):
garage/envs/proxy_env.py:37:        return self._wrapped_env.horizon
garage/envs/box2d/mountain_car_env.py:17:    @autoargs.arg("goal_cart_pos", type=float, help="Goal horizontal position")
garage/algos/ddpg.py:56:                 include_horizon_terminal_transitions=False,
garage/algos/ddpg.py:90:        :param include_horizon_terminal_transitions: whether to include
garage/algos/ddpg.py:91:         transitions with terminal=True because the horizon was reached. This
garage/algos/ddpg.py:126:        self.include_horizon_terminal_transitions = \
garage/algos/ddpg.py:127:            include_horizon_terminal_transitions
garage/algos/ddpg.py:194:                    if self.include_horizon_terminal_transitions:
examples/theano/trpo_gym_pendulum.py:27:        max_path_length=env.horizon,
contrib/bichengcao/examples/trpo_gym_MountainCar-v0.py:7:from garage.envs.util import horizon, spec
contrib/bichengcao/examples/trpo_gym_MountainCar-v0.py:24:        max_path_length=horizon(env),
contrib/bichengcao/examples/trpo_gym_Pendulum-v0.py:6:from garage.envs.util import horizon, spec
contrib/bichengcao/examples/trpo_gym_Pendulum-v0.py:23:        max_path_length=horizon(env),
examples/theano/trpo_gym_cartpole.py:27:        max_path_length=env.horizon,
contrib/bichengcao/examples/trpo_gym_CartPole-v1.py:6:from garage.envs.util import horizon, spec
contrib/bichengcao/examples/trpo_gym_CartPole-v1.py:23:        max_path_length=horizon(env),
contrib/bichengcao/examples/trpo_gym_Acrobot-v1.py:6:from garage.envs.util import horizon, spec
contrib/bichengcao/examples/trpo_gym_Acrobot-v1.py:23:        max_path_length=horizon(env),
contrib/bichengcao/examples/trpo_gym_CartPole-v0.py:6:from garage.envs.util import horizon, spec
contrib/bichengcao/examples/trpo_gym_CartPole-v0.py:23:        max_path_length=horizon(env),
garage/algos/cma_es_lib.py:463:        # learning rates and back-ward time horizons
```